cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py

```diff
@@ -80,22 +85,30 @@ def install():
     command = 'tar -xvzf {0} -C {1}'.format(archive, files_dir)
     hookenv.log(command)
     check_call(split(command))
-    dest_dir = '/usr/local/bin/'
-    # Create a list of components to install.
-    services = ['kube-apiserver',
-                'kube-controller-manager',
-                'kube-scheduler',
-                'kubectl']
-    for service in services:
-        # Install each one of the service binaries in /usr/local/bin.
-        install = 'install -v {0}/{1} {2}'.format(files_dir, service, dest_dir)
-        return_code = call(split(install))
-        if return_code != 0:
-            raise Exception('Unable to install {0}'.format(service))
+
+    apps = [
+        {'name': 'kube-apiserver', 'path': '/usr/local/bin'},
+        {'name': 'kube-controller-manager', 'path': '/usr/local/bin'},
+        {'name': 'kube-scheduler', 'path': '/usr/local/bin'},
+        {'name': 'kubectl', 'path': '/usr/local/bin'},
+    ]
+
+    for app in apps:
+        unpacked = '{}/{}'.format(files_dir, app['name'])
+        app_path = os.path.join(app['path'], app['name'])
+        install = ['install', '-v', '-D', unpacked, app_path]
+        hookenv.log(install)
+        check_call(install)
 
     set_state('kubernetes-master.components.installed')
 
 
+@when('cni.connected')
+@when_not('cni.configured')
+def configure_cni(cni):
```

Comment:
No docstring or comment here to tell me what is going on or why this code exists.